src/experiments/tcpsocket/api.js

```diff
@@ -0,0 +1,91 @@
+"use strict";
+/* exported tcpsocket */
+/* global ExtensionAPI, ChromeUtils, Cu */
+
+const { TCPSocket } = Cu.getGlobalForObject(
+    ChromeUtils.import("resource://gre/modules/Services.jsm")
+);
+
+/**
+ * Concatenate two Uint8Array objects
+ */
+function concatUint8Arrays(a, b) {
+    let newArr = new Uint8Array(a.length + b.length);
+    newArr.set(a, 0);
+    newArr.set(b, a.length);
+    return newArr;
+}
+
+var tcpsocket = class tcpsocket extends ExtensionAPI {
+    getAPI(context) {
+        return {
+            experiments: {
+                tcpsocket: {
+                    /** 
+                     * Send a DNS query stored in buf over a TCP socket to a 
+                     * nameserver addressed by addr
+                     */
+                    async sendDNSQuery(addr, buf) {
+                        let tcp_socket;
+
+                        try {
+                            // Wait until the socket is open before sending data.
+                            // If we get an 'error' event before an 'open' event, 
+                            // throw an ExtensionError.
+                            tcp_socket = new TCPSocket(addr, 53, { binaryType: "arraybuffer" });
+                            let eventType = await new Promise((resolve, reject) => {
+                                tcp_socket.onopen = ((event) => {
```

Comment:
While this works, I would recommend to merge the two promise callbacks.
Call `tcp_socket.send()` from `onopen`. There is no need to check `eventType` nor `readyState`.

Reply:
Addressed in caf03b1 and 62d8337